### PR TITLE
config changes for latest clamav-unofficial-sigs 7.2.5

### PR DESCRIPTION
**Pull request by diroots**
Thanks for submitting a pull request ! Please provide enough information so that others can review your changes. 

For more information, see the `CONTRIBUTING` guide.

## Description

clamav-unofficial-sigs script version loaded in mailserver2 repo is Version: v5.6.2 (2017-03-19)

actual latest version is 7.2.5 : https://github.com/extremeshok/clamav-unofficial-sigs/releases/tag/7.2.5

Fixes https://github.com/mailserver2/mailserver/issues/47

* outdated clamav-unofficial-sigs script

## Type of change

- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Status

- [ ] Ready
- [ ] In development
- [ ] Hold

## Todo List
- [X] Implementation
- [ ] Tests
- [X] Documentation
- [ ] ...

there is a config validation to perform on the os.conf file, as in the mailserver version of the file, only clamd_reload_opt is defined, and in the latest version of this file in the original repo, only clamd_restart_opt is defined. 

also in this file, there is the socket param activated in mailserver version, but not in the original repo version. i did let it enabled for now, to reflect mailserver's setup

## How has this been tested ?

tested on a custom implementation of mailserver stack

- [ ] Test A
- [ ] Test B

### CODE REVIEW

**Review by sknight80 — APPROVED** [review]
Looks good to me! thanks for updating the clamav!

File: README.md
Comment (AndrewSav) [inline README.md:678]: There seems to be an inconsistency:  `etc/clamav/unofficial-sigs` here vs `etc/clamav-unofficial-sigs` elsewhere, why was this changed?

**Comment by AndrewSav** [comment]
It looks like we are copying`clamav-unofficial-sigs.sh` and `master.conf` from <https://github.com/extremeshok/clamav-unofficial-sigs>. What do you guys thing of the following suggestion:

- Pull the file during the docker build instead of copying
- Identify it for pulling by commit hash / tag
- Authenticate it with checksum(s) that we will store in our repo and re-calculate each time we change the commit hash / tag to pull from
- Perhaps create a simple shell script for getting the checksum(s) we can paste on update

This will make obvious that this is an external dependency, will save time checking that no errors were made copying files from another repo and we will have more or less straight forward process of updating it in future.

@diroots @sknight80 @SaraSmiseth opinions?

**Review by AndrewSav — CHANGES_REQUESTED** [review]

**Comment by sknight80** [comment]
Sounds good to me. How we will collect the right hash tag and checksum?  Do you think we can automate it?

**Comment by AndrewSav** [comment]
> How we will collect the right hash tag and checksum? Do you think we can automate it?

That's what I mean by `Perhaps create a simple shell script for getting the checksum(s) we can paste on update`. We will always need a human to test that the latest of that other repo works for us. When time comes to update our repo from theirs we select commit hash / tag on the other repo and run that script which will get the checksums, Now we can put the new commit hash / tag and the new checksums in our repo.

This should not be difficult, but it requires a bit of time to implement. 

We can get even fancier with GitHub actions, and detect that if someone committed a new commit hash / tag from the other repo, we can go get the checksum and author another commit with the checksums - I know this should be possible I did something similar elsewhere. I would not burden our contributor with this part though and do not have time for that right now myself. We can always improve this later once we get started.

File: README.md
Comment (diroots) [reply README.md:678]: i just noticed it had changed when launching the clamav-unofficial-sigs.sh script, it seems a long tech debt, as even in the 5.6.2 official release it's already `/etc/clamav-unofficial-sigs` see https://github.com/extremeshok/clamav-unofficial-sigs/releases/tag/5.6.2 , but https://github.com/hardware/mailserver/blob/master/rootfs/usr/local/bin/clamav-unofficial-sigs.sh#L1293 :'( 

even is same script info : 

```
# Script Info
script_version="5.6.2"
script_version_date="2017-03-19"
minimum_required_config_version="72"
minimum_yara_clamav_version="0.99"
```

Comment (AndrewSav) [reply README.md:678]: Ah, so it was changed on their side, fair enough. Could you please update the readme.md to reflect the new path then? Many thanks!

Comment (diroots) [reply README.md]: it seems changed now in "Files changed" can you confirm and resolve?

**Comment by diroots** [comment]
> It looks like we are copying`clamav-unofficial-sigs.sh` and `master.conf` from https://github.com/extremeshok/clamav-unofficial-sigs. What do you guys thing of the following suggestion:
> 
>     * Pull the file during the docker build instead of copying
> 
>     * Identify it for pulling by commit hash / tag
> 
>     * Authenticate it with checksum(s) that we will store in our repo and re-calculate each time we change the commit hash / tag to pull from
> 
>     * Perhaps create a simple shell script for getting the checksum(s) we can paste on update
> 
> 
> This will make obvious that this is an external dependency, will save time checking that no errors were made copying files from another repo and we will have more or less straight forward process of updating it in future.
> 
> @diroots @sknight80 @SaraSmiseth opinions?

it should be interesting to get files based on their latest tag, ie : https://github.com/extremeshok/clamav-unofficial-sigs/tree/7.2.5

and then grab directly all 3 files from there : 
- clamav-unofficial-sigs.sh
- config/master.conf
- config/os/os.debian.conf (this one needing renaming to os.conf and also the specific changes for mailserver setup, ie. s6svc and socket) 

see https://github.com/extremeshok/clamav-unofficial-sigs/blob/7.2.5/guides/ubuntu-debian.md#install

File: README.md
Comment (AndrewSav) [reply README.md]: Thank you, I just searched the repo for `clamav/unofficial` and there are two more occurrences, one in Makefile and one in os.conf, can you check it out please?

**Comment by AndrewSav** [comment]
> it should be interesting to get files based on their latest tag, ie : https://github.com/extremeshok/clamav-unofficial-sigs/tree/7.2.5
> 
> and then grab directly all 3 files from there :
> 
> * clamav-unofficial-sigs.sh
> * config/master.conf
> * config/os/os.debian.conf (this one needing renaming to os.conf and also the specific changes for mailserver setup, ie. s6svc and socket)
> 
> see https://github.com/extremeshok/clamav-unofficial-sigs/blob/7.2.5/guides/ubuntu-debian.md#install

Sounds reasonable. I understand that os.debian.conf require some modifications though, so cannot be grabbed as it, how do we resolve it? Also @diroots would you be willing to implement this?

File: README.md
Comment (diroots) [reply README.md]: @AndrewSav the one in Makefile seems accurate, as is it the link to test folder : 

`		-v "`pwd`/test/share/clamav/unofficial-sigs/user.conf":/var/mail/clamav-unofficial-sigs/user.conf \`


for the one in os.conf that are the specific points i wanted you to check/validate/correct as it relates to the todo list i did write on the desc of the MR

Comment (AndrewSav) [reply README.md]: @diroots hello, thank you for this. For `Makefile`: don't you think that for consistency we should rename the test folder too?

For `os.conf` can you share a bit more of why you are not sure? To me it seems that leaving in there a path that was changed everywhere else does not make sense, although I could miss something. Why do you think it might make sense?

Comment (diroots) [reply README.md]: @AndrewSav for Makefile, could be good to rename the testfolder too. will it break any test? 

for os.conf : the references to old path are comments I did let on file for you, maintainers of  mailserver app, to review. as i'm not sure of what should be final definition, let me paste here to exchange on it : 


```
# ### https://github.com/mailserver2/mailserver specifics : ###
# taken from https://github.com/mailserver2/mailserver/blob/master/rootfs/etc/clamav/unofficial-sigs/os.conf#L7
# as service handler is s6-supervise

clamd_reload_opt="s6-svc -2 /services/clamd"
# this is in case the needed/expected keyword is now restart and not reload
clamd_restart_opt="s6-svc -2 /services/clamd"


# this is enabled here https://github.com/mailserver2/mailserver/blob/master/rootfs/etc/clamav/unofficial-sigs/os.conf#L6
clamd_socket="/run/clamav/clamd.ctl"
```

in your old os.conf file, the clamd_reload_opt was used, and modified to use s6-svc with signal 2 which is a SIGUSR2 signal. SIGUSR2 performs a Reload the signature databases. as per https://manpages.debian.org/bullseye/clamav-daemon/clamd.8.en.html#SIGUSR2


in latest os.debian.conf on the clamav-unofficiel-sigs project, https://github.com/extremeshok/clamav-unofficial-sigs/blob/master/config/os/os.debian.conf#L33 they do define a clamd_restart_opt value.

what is the s6svc command to perform a proper clamav restart? i did try s6-svc -r /services/clamd but clamd process do not respawn correctly. (taken from https://skarnet.org/software/s6/s6-svc.html) 

this is after a `s6-svc -r /services/clamd` or even after a `s6-svc -d /services/clamd && s6-svc -u /services/clamd`
Tue Nov 29 14:37:09 2022 -> --- Stopped at Tue Nov 29 14:37:09 2022
Tue Nov 29 14:37:09 2022 -> Socket file removed.

i need to force a single `s6-svc -u /services/clamd` to see clamd restart correctly
Tue Nov 29 14:38:55 2022 -> +++ Started at Tue Nov 29 14:38:55 2022
Tue Nov 29 14:38:55 2022 -> Received 0 file descriptor(s) from systemd.
Tue Nov 29 14:38:55 2022 -> clamd daemon 0.103.3 (OS: linux-gnu, ARCH: x86_64, CPU: x86_64)


also for last comment about the socket. i'm not sure if the socket is linked to clamav as i don't find any reference to it in clamav config files,... so do we need to keep it? 

and last info about os.conf 

in your original version, there were these 2 additional params : 

```
min_sleep_time="30"
max_sleep_time="60"
```

do we need to keep them or are these useless?

Comment (AndrewSav) [reply README.md]: @diroots 
> for Makefile, could be good to rename the testfolder too. will it break any test?

That would be good to check out, this is what the tests are for: we need to run them after any changes to make sure they do not break, I'd say it should be a pre-requisite for a successful merge. I would run them locally (a gotcha is that they cannot be run as root, I wasted a fair bit of time before realising that), but the PR pipeline can run them too, I approved it now for a first time contributor, all subsequent ones should not need an approval. When you ensure that tests work after that change please report back, or if they break and you cannot figure out why, let me know and I'll have a look.

> what is the s6svc command to perform a proper clamav restart? i did try s6-svc -r /services/clamd but clamd process do not respawn correctly. 

It's a tricky one. There is a [finish script](https://github.com/mailserver2/mailserver/blob/71688532934217c117d79e80b89b2bd36eeadefc/rootfs/usr/local/bin/setup.sh#L814) that is added to each service. It is intended to to stop restarting services if they restart too many times. but the side effect is that if a service finish with 0 exit code, `s6-svc -d` is called which will then prevent for the service to come up. Since the same finish script is used for all the services, not just for clamd I would imagine that the intention was that some of those are _meant_ to finish eventually and permanently. I would question why they are excuted with s6 then, but I did not design this so I can only guess. 

One workaround that I came up with, although it is not ideal, is to create a `down-signal` file in `/services/clamd` which contains a single line saying `SIGKILL`. It will make ` s6-svc -r /services/clamd` send `SIGKILL` instead of `SIGTERM` so the `finish` script will receive `256 9` instead of exit code of 0, and restart will proceed as planned. This is not ideal, because SIGTERM is obviously a more gentle approach, but without good understanding why exactly we have the `finish` script as is for all the services it may require more testing to come up with a different solution. If no one has a better idea and there are no visible drawbacks `SIGTERM` may work for us.

> also for last comment about the socket. i'm not sure if the socket is linked to clamav as i don't find any reference to it in clamav config files,... so do we need to keep it?

Sorry, which socket is that, can you give some more info?

>do we need to keep them or are these useless?

Why do you think they are useless? the are different from [the defaults](https://github.com/extremeshok/clamav-unofficial-sigs/blob/4a5b8f1ef64a6f1f744c98b605d7ed83ce5c7453/config/master.conf#L572) so to me it looks like they serve some purpose? I do not mind removing them since they do not seem to matter much, the defaults seem sensible, but on the other hand some one added them there with some purpose which we do not know, so may be it is safer to leave them in.

**Comment by AndrewSav** [comment]
For what it's worth I had a conversation on skarnet irc with regards to the restart issue. They informed me that using `s6-svc -d` in a finish script across the board as we do is misguided. It can only make sense in a rare case when supervise should not bring back the process after it finished, e.g it's not a daemon. Since most of our processes are daemons, this finish script is not very suitable for them, and this is something that probably could be improved in future. For me the main snag is to identify which process _can_ legitimately finish.

File: README.md
Comment (diroots) [reply README.md]: > Sorry, which socket is that, can you give some more info?

this socket : 
https://github.com/mailserver2/mailserver/blob/master/rootfs/etc/clamav/unofficial-sigs/os.conf#L6

> I do not mind removing them since they do not seem to matter much, the defaults seem sensible, but on the other hand some one added them there with some purpose which we do not know, so may be it is safer to leave them in

these timeouts help during the time randomization of signatures updates ran by cron. 
in original, it let between 1 and 10 minutes. in this setup, it narrow this between 30s and one minut. 

> # Enable or disable download time randomization.  This allows the script to
> # be executed via cron, but the actual database file checking will pause
> # for a random number of seconds between the "min" and "max" time settings
> # specified below.  This helps to more evenly distribute load on the host
> # download sites.

bigger timeframe helps to more evenly distribute load on the host download sites. so narrowing the sleep times generate more requests more often on download sites. 
moreover signatures do not need to be updated so often. and some signature host sites warn you can be blacklisted if you do request too often. i would let default values there. (it's what i did on our setup, and it do not seems to cause problems as per the logs.

Comment (diroots) [reply README.md]: SIGKILL is arsh and can lead to ongoing clamd executions being stopped while running. 

from there https://github.com/extremeshok/clamav-unofficial-sigs/blob/master/clamav-unofficial-sigs.sh#L1336 it seems restart is fired only if reload is not fully working. in your previous implementations, you were relying only on reload only, as per your previous os.conf file. should we stay as is ? whatever the script will log if reload fails and restart not implemented

**Comment by diroots** [comment]
> Since most of our processes are daemons, this finish script is not very suitable for them, and this is something that probably could be improved in future.

https://github.com/mailserver2/mailserver/blob/71688532934217c117d79e80b89b2bd36eeadefc/rootfs/usr/local/bin/setup.sh#L731 there "Foreground" is set to true, so clamd might not be demonized

**Comment by diroots** [comment]
> Sounds reasonable. I understand that os.debian.conf require some modifications though, so cannot be grabbed as it, how do we resolve it? Also @diroots would you be willing to implement this?

in the MR i proposed to get the os.conf from their os.debian.conf but in the end modified it to make it similar to your os.conf based on s6,... so could be not worth it to use the debian based one proposed, as we nearly do not keep any params from it.

File: README.md
Comment (sknight80) [reply README.md]: @AndrewSav if you think we can rethink the startup process to improve it in a separate PR?

**Comment by sknight80** [comment]
> > Sounds reasonable. I understand that os.debian.conf require some modifications though, so cannot be grabbed as it, how do we resolve it?
> 
> in the MR i proposed to get the os.conf from their os.debian.conf but in the end I modified it to make it similar to your os.conf based on s6,... so could be not worth it to use the debian based one proposed, as we nearly do not keep any params from it.

I am okay with keeping the os.conf file that we had before.  But I am also fine, to grab the os.debian.conf and write a script that is going to update to our setup (this potentially more work)

**Comment by sknight80** [comment]
@AndrewSav are you okay with this PR?  If so, can we merge it?

**Comment by AndrewSav** [comment]
@sknight80 Sorry, I did not realise it was waiting on me which it seems it is. I will get back to you soon, I think that some of the changes that we discussed above are better be done in scope of this PR, but I cannot tell you which one just at this second, I'll try to review and get back before Monday.

**Comment by AndrewSav** [comment]
> there "Foreground" is set to true, so clamd might not be demonized

I think conceptually the are still demons, we use Foregraound simply because we are using s6 for supervising them. As such I think my earlier comment you are replying to still stands.

File: README.md
Comment (AndrewSav) [reply README.md]: >also for last comment about the socket. i'm not sure if the socket is linked to clamav as i don't find any reference to it in clamav config files,... so do we need to keep it?
>this socket : https://github.com/mailserver2/mailserver/blob/master/rootfs/etc/clamav/unofficial-sigs/os.conf#L6

It is mentioned here: https://github.com/mailserver2/mailserver/blob/master/rootfs/etc/clamav/unofficial-sigs/master.conf#L423 and if I'm not mistaken it is still there with the latest version in the clamav unofficial sigs repo. Do you think they forgot to remove it from there by mistake and it is not longer used? If so may be worth opening issue with them to clarify. Otherwise for the purpose of merging the PR I'd probably leave it there on the off chance we are missing it's usage.

> bigger timeframe helps to more evenly distribute load on the host download sites. so narrowing the sleep times generate more requests more often on download sites.
>moreover signatures do not need to be updated so often. and some signature host sites warn you can be blacklisted if you do request too often. i would let default values there. (it's what i did on our setup, and it do not seems to cause problems as per the logs.

Ok, that's sounds reasonable, let's remove them then.

> from there https://github.com/extremeshok/clamav-unofficial-sigs/blob/master/clamav-unofficial-sigs.sh#L1336 it seems restart is fired only if reload is not fully working. in your previous implementations, you were relying only on reload only, as per your previous os.conf file. should we stay as is ? whatever the script will log if reload fails and restart not implemented

If reload works, then leaving as it is is the preferred option, but from what you wrote earlier I thought that they no longer using reload and using restart instead. If that's not the case and reload still works as it used to, let's stick with how it used to work.

---

In order to merge the PR we want: 
- change to the test folder path for clamav to make it consistent across the board and fix broken tests if any
- remove the `clamd_restart_opt` option since it's not working (if reload will still work - otherwise we need to think of something else)
- remove `min_sleep_time` and `max_sleep_time` since it looks like the defaults are better

**Comment by AndrewSav** [comment]
@sknight80 @diroots I left my comments in the change thread above. Thanks! Sorry for the delay I have no idea how this has fallen through the cracks I was somehow sure it was not on my plate.

File: README.md
Comment (diroots) [reply README.md]: ok @AndrewSav see https://github.com/diroots/mailserver/commit/12fe85b896879d5a45d709240e291126184a7a05

**Comment by AndrewSav** [comment]
@diroots can you rebase on master please?

**Comment by AndrewSav** [comment]
I built [mailserver2/mailserver:test_cus](https://hub.docker.com/layers/mailserver2/mailserver/test_cus/images/sha256-d094bd3d5d74881d6c299ed1362e500e2ec49602aef4a403e33e51a23114af9b?context=explore) from this PR, I rebased it locally. It seems to run fine. I think once we rebase it and see that the tests are green we will be ready to merge. Thank you again for all your hard work, and apologies for the very slow pace from our side.

**Comment by diroots** [comment]
> @diroots can you rebase on master please?

done @AndrewSav

**Review by AndrewSav — APPROVED** [review]

**Comment by AndrewSav** [comment]
@diroots this is in 1.1.14 image now. Thank you again for making this happen!

**Pull request merged**
